File: back/services/alerta_masiva_service.py
from typing import List, Dict, Optional
from sqlalchemy.orm import Session
from sqlalchemy import select, insert
from config.db import engine as shared_engine
from models.TipoReporte import TipoReporte
from models.Reporte import Reporte
from models.Corredor import Corredor
from models.Ruta import Ruta
from models.Paradero import Paradero
from datetime import datetime, timezone
from models.UsuarioBase import UsuarioBase
from config.firebase import get_firebase_admin

class AlertaMasivaService:
    def __init__(self, engine=None):
        self.engine = engine or shared_engine
        if self.engine is None:
            raise RuntimeError("No hay engine de DB disponible en AlertaMasivaService.")

    def obtener_datos_formulario(self) -> Dict:
        """
        Obtiene todos los datos necesarios para poblar los dropdowns del formulario:
        - Corredores
        - Rutas
        - Paraderos
        """
        try:
            with Session(self.engine) as session:
                # Obtener corredores
                stmt_corredores = select(Corredor)
                corredores = session.execute(stmt_corredores).scalars().all()
                
                # Obtener rutas
                stmt_rutas = select(Ruta)
                rutas = session.execute(stmt_rutas).scalars().all()
                
                # Obtener paraderos
                stmt_paraderos = select(Paradero)
                paraderos = session.execute(stmt_paraderos).scalars().all()
                
                return {
                    "corredores": [
                        {
                            "id_corredor": c.id_corredor,
                            "nombre": c.nombre
                        }
                        for c in corredores
                    ],
                    "rutas": [
                        {
                            "id_ruta": r.id_ruta,
                            "codigo": r.codigo,
                            "nombre": r.nombre
                        }
                        for r in rutas
                    ],
                    "paraderos": [
                        {
                            "id_paradero": p.id_paradero,
                            "nombre": p.nombre
                        }
                        for p in paraderos
                    ]
                }
        except Exception as e:
            print(f"[ERROR] AlertaMasivaService.obtener_datos_formulario: {e}")
            raise

    def crear_alerta_masiva(self, payload: Dict) -> Dict:
        """
        Crea un nuevo reporte tipo "Otro" (id_tipo_reporte = 4) para alertas masivas.
        Guarda todos los campos del reporte en la base de datos.
        Si se especifica, envía una notificación masiva a un tema.
        """
        try:
            with Session(self.engine) as session:
                # Preparar los datos del reporte
                datos_reporte = {
                    "fecha": datetime.now(timezone.utc),
                    "descripcion": payload.get("descripcion"),
                    "id_emisor": payload.get("id_emisor"),
                    "id_tipo_reporte": 4,  # Tipo "Otro" para alertas masivas
                    "id_corredor_afectado": payload.get("id_corredor_afectado"),
                    "es_critica": payload.get("es_critica", False),
                    "requiere_intervencion": payload.get("requiere_intervencion", False),
                    "id_ruta_afectada": payload.get("id_ruta_afectada"),
                    "id_paradero_inicial": payload.get("id_paradero_inicial"),
                    "id_paradero_final": payload.get("id_paradero_final"),
                    "tiempo_retraso_min": payload.get("tiempo_retraso_min")
                }
                
                # Insertar el nuevo reporte
                stmt = insert(Reporte).values(**datos_reporte).returning(Reporte)
                result = session.execute(stmt)
                nuevo_reporte = result.scalar_one()
                
                session.commit()

                # Si se solicita, enviar notificación al tema "all_users"
                if payload.get("send_notification"):
                    try:
                        firebase_admin = get_firebase_admin()
                        firebase_admin.send_to_topic(
                            topic="all_users",
                            title="Alerta General",
                            body=payload.get("descripcion")
                        )
                        print("✅ Notificación de alerta masiva enviada al tema 'all_users'")
                    except Exception as e:
                        # Log del error pero no fallar la solicitud, ya que la alerta fue creada
                        print(f"❌ [ERROR] Al enviar notificación de alerta masiva al tema: {e}")
                
                return {
                    "id_reporte": nuevo_reporte.id_reporte,
                    "fecha": nuevo_reporte.fecha.isoformat(),
                    "descripcion": nuevo_reporte.descripcion,
                    "id_emisor": nuevo_reporte.id_emisor,
                    "id_tipo_reporte": nuevo_reporte.id_tipo_reporte,
                    "es_critica": nuevo_reporte.es_critica,
                    "requiere_intervencion": nuevo_reporte.requiere_intervencion,
                    "mensaje": "Alerta masiva creada exitosamente"
                }
        except Exception as e:
            print(f"[ERROR] AlertaMasivaService.crear_alerta_masiva: {e}")
            raise

File: front/services/alertaMasivaService.ts
import AsyncStorage from "@react-native-async-storage/async-storage";
import { API_URL } from './userService';

interface DatosFormulario {
    corredores: Array<{ id_corredor: number; nombre: string }>;
    rutas: Array<{ id_ruta: number; codigo: string; nombre: string }>;
    paraderos: Array<{ id_paradero: number; nombre: string }>;
}

interface AlertaMasivaPayload {
    descripcion: string;
    id_corredor_afectado?: number;
    es_critica?: boolean;
    requiere_intervencion?: boolean;
    id_ruta_afectada?: number;
    id_paradero_inicial?: number;
    id_paradero_final?: number;
    tiempo_retraso_min?: number;
}

/**
 * Obtiene los datos necesarios para el formulario (corredores, rutas, paraderos)
 */
export async function obtenerDatosFormulario(): Promise<DatosFormulario> {
    try {
        console.log('🔍 [AlertaMasiva] Cargando datos del formulario...');
        console.log('🌐 [AlertaMasiva] API_URL:', API_URL);
        
        // Obtener corredores, rutas y paraderos de sus endpoints respectivos
        const [corredoresRes, rutasRes, paraderosRes] = await Promise.all([
            fetch(`${API_URL}/corredor/`, { method: 'GET' }),
            fetch(`${API_URL}/ruta/obtenerRutas`, { method: 'GET' }),
            fetch(`${API_URL}/paradero/`, { method: 'GET' })
        ]);

        console.log('📡 [AlertaMasiva] Status codes:', {
            corredores: corredoresRes.status,
            rutas: rutasRes.status,
            paraderos: paraderosRes.status
        });

        if (!corredoresRes.ok || !rutasRes.ok || !paraderosRes.ok) {
            throw new Error('Error al obtener datos del formulario');
        }

        const [corredoresData, rutasData, paraderosData] = await Promise.all([
            corredoresRes.json(),
            rutasRes.json(),
            paraderosRes.json()
        ]);

        console.log('✅ [AlertaMasiva] Datos obtenidos:', {
            corredores: corredoresData.length,
            rutas: rutasData.length,
            paraderos: paraderosData.length
        });

        console.log('📦 [AlertaMasiva] Datos raw:', {
            corredoresData,
            rutasData,
            paraderosData
        });

        return {
            corredores: corredoresData.map((c: any) => ({
                id_corredor: c.id_corredor,
                capacidad_max: c.capacidad_max,
                ubicacion_lat: c.ubicacion_lat,
                ubicacion_lng: c.ubicacion_lng,
                estado: c.estado
            })),
            rutas: rutasData.map((r: any) => ({
                id_ruta: r.id_ruta,
                nombre: r.nombre
            })),
            paraderos: paraderosData.map((p: any) => ({
                id_paradero: p.id_paradero,
                nombre: p.nombre
            }))
        };
    } catch (error) {
        console.error('Error en obtenerDatosFormulario:', error);
        throw error;
    }
}

/**
 * Crea una nueva alerta masiva (reporte tipo "Otro")
 */
export async function crearAlertaMasiva(payload: AlertaMasivaPayload): Promise<any> {
    try {
        console.log('📝 [AlertaMasiva] Guardando alerta...');
        console.log('📋 [AlertaMasiva] Payload:', payload);
        
        // Obtener el token de autenticación
        const token = await AsyncStorage.getItem('token');
        
        if (!token) {
            throw new Error('No hay sesión activa');
        }

        console.log('🔑 [AlertaMasiva] Token obtenido');

        const url = `${API_URL}/alertas-masivas/enviar/`;
        console.log('📡 [AlertaMasiva] URL:', url);
        
        const headers: Record<string, string> = {
            'Content-Type': 'application/json',
            'Authorization': `Bearer ${token}`
        };

        const response = await fetch(url, {
            method: 'POST',
            headers,
            body: JSON.stringify(payload),
        });

        const responseText = await response.text();
        let data: any;

        try {
            data = JSON.parse(responseText);
        } catch {
            data = responseText;
        }

        if (!response.ok) {
            throw new Error(
                typeof data === 'object' && data.detail 
                    ? data.detail 
                    : `Error al crear alerta masiva: ${response.status}`
            );
        }

        return data;
    } catch (error) {
        console.error('Error en crearAlertaMasiva:', error);
        throw error;
    }
}

/**
 * Envía una notificación push masiva a todos los usuarios.
 */
export async function enviarNotificacionAlertaMasiva(descripcion: string): Promise<any> {
    try {
        console.log('🔔 [AlertaMasiva] Enviando notificación...');
        console.log('📋 [AlertaMasiva] Descripción:', descripcion);
        
        const token = await AsyncStorage.getItem('token');
        
        if (!token) {
            throw new Error('No hay sesión activa');
        }

        const url = `${API_URL}/alertas-masivas/enviar-notificacion/`;
        console.log('📡 [AlertaMasiva] URL de notificación:', url);
        
        const headers: Record<string, string> = {
            'Content-Type': 'application/json',
            'Authorization': `Bearer ${token}` // Endpoint requiere autenticación
        };

        const response = await fetch(url, {
            method: 'POST',
            headers,
            body: JSON.stringify({ descripcion: descripcion }),
        });

        const responseText = await response.text();
        let data: any;

        try {
            data = JSON.parse(responseText);
        } catch {
            data = responseText;
        }

        if (!response.ok) {
            throw new Error(
                typeof data === 'object' && data.detail 
                    ? data.detail 
                    : `Error al enviar notificación masiva: ${response.status}`
            );
        }

        return data;
    } catch (error) {
        console.error('Error en enviarNotificacionAlertaMasiva:', error);
        throw error;
    }
}

export default {
    obtenerDatosFormulario,
    crearAlertaMasiva,
    enviarNotificacionAlertaMasiva
};
